### PR TITLE
Use LogIndex for log index

### PR DESCRIPTION
**Pull request by markcty**
Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)
Use `LogIndex`(u64) for log index. We previously use `usize` for log index. `usize` may have different size on different machines, which may result in unexpected bugs.

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: curp/src/server/raw_curp/log.rs [hunk 1]
```diff
@@ -52,17 +54,17 @@ impl<C: 'static + Command> Log<C> {
             entries, // a fake log[0]
             commit_index: 0,
             last_applied: 0,
-            base_index: 1,
+            base_index: 0,
```
Comment (Phoenix500526) [inline curp/src/server/raw_curp/log.rs:57]: Why don't we need fake log[0] entry now? Is there any use for it other than simplify the calculation?

Comment (markcty) [reply curp/src/server/raw_curp/log.rs:57]: No, we still have a fake log[0]. I changed the meaning of base_index(see the comments). I think it's more straight forward this way.

**Review by Phoenix500526 — APPROVED** [review]

File: curp/src/server/raw_curp/log.rs [hunk 2]
```diff
@@ -165,18 +166,19 @@ impl<C: 'static + Command> Log<C> {
...
             let entry = self.get(i - 1).unwrap_or_else(|| {
                 unreachable!(
-                    "system corrupted, get log[{i}] when we only have {} log entries",
+                    "system corrupted, get log[{}] when we only have {} log entries",
+                    i - 1,
                     self.last_log_index()
                 )
             });
-            (entry.term, entry.index)
+            (entry.index, entry.term)
```
Comment (themanforfree) [inline curp/src/server/raw_curp/log.rs:181]: Why reverse the order here, I see term in front of index in other places

File: curp/src/server/raw_curp/mod.rs [hunk 3]
```diff
@@ -317,11 +317,11 @@ impl<C: 'static + Command> RawCurp<C> {
...
         term: u64,
         leader_id: String,
-        prev_log_index: usize,
+        prev_log_index: LogIndex,
         prev_log_term: u64,
         entries: Vec<LogEntry<C>>,
-        leader_commit: usize,
-    ) -> Result<u64, (u64, usize)> {
+        leader_commit: LogIndex,
+    ) -> Result<u64, (u64, LogIndex)> {
```
Comment (themanforfree) [inline curp/src/server/raw_curp/mod.rs:324]: term in front of index, or reverse? please unify the format

File: curp/src/server/raw_curp/log.rs (hunk 2 above)
Comment (markcty) [reply curp/src/server/raw_curp/log.rs:181]: To unify with the format in other places

**Review by themanforfree — APPROVED** [review]

**Pull request merged**